src/service/position-observer-impl.js

```diff
@@ -0,0 +1,413 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use baseInstance file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {registerServiceBuilderForDoc} from '../service';
+import {viewportForDoc, vsyncFor} from '../services';
+import {getMode} from '../mode';
+import {dev} from '../log';
+import {
+  moveLayoutRect,
+  layoutRectEquals,
+  layoutRectsOverlap,
+  layoutRectFromDomRect,
+  layoutRectLtwh,
+} from '../layout-rect';
+import {IframeMessagingClient} from '../../3p/iframe-messaging-client';
+import {isExperimentOn} from '../../src/experiments';
+import {MessageType} from '../../src/3p-frame-messaging';
+
+
+/** @const @private */
+const TAG = 'POSITION_OBSERVER';
+
+/**
+ * The positionObserver returned position value which includes the position rect
+ * relative to viewport. And viewport rect which always has top 0, left 0, and
+ * viewport width and height.
+ * @typedef {{
+ *  positionRect: ?../layout-rect.LayoutRectDef,
+ *  viewportRect: !../layout-rect.LayoutRectDef,
+ * }}
+ */
+export let PositionInViewportEntryDef;
+
+/** @enum {number} */
+export const PositionObserverFidelity = {
+  HIGH: 1,
+  LOW: 0,
+};
+
+/** @const @private */
+const LOW_FEDELITY_FRAME_COUNT = 2;
+
+class AbstractPositionObserver {
+
+  /** @param {!./ampdoc-impl.AmpDoc} ampdoc */
+  constructor(ampdoc) {
+    /** @const {!./ampdoc-impl.AmpDoc} */
+    this.ampdoc_ = ampdoc;
+
+    /** @private {!Array<!Object>} */
+    this.entries_ = [];
+
+    /** @private {!./vsync-impl.Vsync} */
+    this.vsync_ = vsyncFor(ampdoc.win);
+
+    /** @private {!./viewport-impl.Viewport} */
+    this.viewport_ = viewportForDoc(ampdoc);
+
+  }
+
+  /**
+   * @param {!Element} element
+   * @param {PositionObserverFidelity} fidelity
+   * @param {function(PositionInViewportEntryDef)} handler
+   */
+  observe(element, fidelity, handler) {
+    // make entry into a class
+    const entry = {
```

Comment:
Entry should be a class, so that `#trigger` doesn't need to be redefined on every instance.